File: examples/08-ai-ml/java/ml_pipeline_workflow.java
// An ML pipeline as a Workflow sequencing Activities: per-step retry policy, a
// quality gate written in code, heartbeat-based resume on the long training step,
// and idempotency via the deterministic workflowId passed into Activities. The
// Day-1 Airflow-vs-Temporal contrast, applied to model training.
@WorkflowInterface
interface MLPipelineWorkflow {
  @WorkflowMethod
  String run(String datasetUri, double accuracyThreshold);

  @SignalMethod
  void approveDeploy();
}

class MLPipelineWorkflowImpl implements MLPipelineWorkflow {
  // Flaky source: retry generously.
  private final MLActivities ingestStub =
      Workflow.newActivityStub(
          MLActivities.class,
          ActivityOptions.newBuilder()
              .setStartToCloseTimeout(Duration.ofMinutes(30))
              .setRetryOptions(RetryOptions.newBuilder().setMaximumAttempts(10).build())
              .build());

  // Long GPU job: heartbeats let it resume from the last checkpoint on retry
  // instead of restarting from epoch 0. Cap attempts so a broken run doesn't
  // loop on GPUs.
  private final MLActivities trainStub =
      Workflow.newActivityStub(
          MLActivities.class,
          ActivityOptions.newBuilder()
              .setStartToCloseTimeout(Duration.ofHours(12))
              .setHeartbeatTimeout(Duration.ofMinutes(5))
              .setRetryOptions(RetryOptions.newBuilder().setMaximumAttempts(3).build())
              .build());

  private boolean approved = false;

  @Override
  public String run(String datasetUri, double accuracyThreshold) {
    String runId = Workflow.getInfo().getWorkflowId(); // deterministic → idempotent

    String raw = ingestStub.ingest(datasetUri);
    String features = ingestStub.preprocess(raw);
    String modelUri = trainStub.train(features, runId);
    Metrics metrics = trainStub.evaluate(modelUri);

    // Quality gate, in code: auto-deploy if clearly good, otherwise wait for a
    // human approval Signal before promoting a borderline model.
    if (metrics.accuracy() < accuracyThreshold) {
      Workflow.await(() -> approved);
    }
    return trainStub.deploy(modelUri, runId);
  }

  @Override
  public void approveDeploy() {
    approved = true;
  }
}

File: examples/08-ai-ml/go/ml_pipeline_workflow.go
package aiml

import (
	"time"

	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

// An ML pipeline as a Workflow sequencing Activities: per-step retry policy, a
// quality gate written in code, heartbeat-based resume on the long training step,
// and idempotency via the deterministic WorkflowID passed into Activities. The
// Day-1 Airflow-vs-Temporal contrast, applied to model training.
func MLPipelineWorkflow(ctx workflow.Context, datasetURI string, accuracyThreshold float64) (string, error) {
	runID := workflow.GetInfo(ctx).WorkflowExecution.ID // deterministic → idempotent

	approved := false
	workflow.Go(ctx, func(ctx workflow.Context) {
		workflow.GetSignalChannel(ctx, "approveDeploy").Receive(ctx, nil)
		approved = true
	})

	// Flaky source: retry generously.
	ingestCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Minute,
		RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: 10},
	})
	var raw, features string
	if err := workflow.ExecuteActivity(ingestCtx, Ingest, datasetURI).Get(ctx, &raw); err != nil {
		return "", err
	}
	if err := workflow.ExecuteActivity(ingestCtx, Preprocess, raw).Get(ctx, &features); err != nil {
		return "", err
	}

	// Long GPU job: heartbeats let it resume from the last checkpoint on retry
	// instead of restarting from epoch 0. Cap attempts so a broken run doesn't
	// loop on GPUs.
	trainCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 12 * time.Hour,
		HeartbeatTimeout:    5 * time.Minute,
		RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: 3},
	})
	var modelURI string
	if err := workflow.ExecuteActivity(trainCtx, Train, features, runID).Get(ctx, &modelURI); err != nil {
		return "", err
	}

	evalCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 20 * time.Minute,
	})
	var metrics Metrics
	if err := workflow.ExecuteActivity(evalCtx, Evaluate, modelURI).Get(ctx, &metrics); err != nil {
		return "", err
	}

	// Quality gate, in code: auto-deploy if clearly good, otherwise wait for a
	// human approval Signal before promoting a borderline model.
	if metrics.Accuracy < accuracyThreshold {
		_ = workflow.Await(ctx, func() bool { return approved })
	}

	var endpoint string
	if err := workflow.ExecuteActivity(evalCtx, Deploy, modelURI, runID).Get(ctx, &endpoint); err != nil {
		return "", err
	}
	return endpoint, nil
}
